// ngx-pinch-zoom/esm5/lib/ivypinch.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ivypinch.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Touches } from './touches';
/**
 * @record
 */
export function IvyPinchProperties() { }
if (false) {
    /** @type {?} */
    IvyPinchProperties.prototype.element;
    /** @type {?|undefined} */
    IvyPinchProperties.prototype.key;
    /** @type {?|undefined} */
    IvyPinchProperties.prototype.doubleTap;
    /** @type {?|undefined} */
    IvyPinchProperties.prototype.doubleTapScale;
    /** @type {?|undefined} */
    IvyPinchProperties.prototype.zoomControlScale;
    /** @type {?|undefined} */
    IvyPinchProperties.prototype.transitionDuration;
    /** @type {?|undefined} */
    IvyPinchProperties.prototype.autoZoomOut;
    /** @type {?|undefined} */
    IvyPinchProperties.prototype.limitZoom;
    /** @type {?|undefined} */
    IvyPinchProperties.prototype.disablePan;
    /** @type {?|undefined} */
    IvyPinchProperties.prototype.limitPan;
    /** @type {?|undefined} */
    IvyPinchProperties.prototype.minPanScale;
    /** @type {?|undefined} */
    IvyPinchProperties.prototype.minScale;
    /** @type {?|undefined} */
    IvyPinchProperties.prototype.eventHandler;
    /** @type {?|undefined} */
    IvyPinchProperties.prototype.listeners;
    /** @type {?|undefined} */
    IvyPinchProperties.prototype.wheel;
    /** @type {?|undefined} */
    IvyPinchProperties.prototype.autoHeight;
    /** @type {?|undefined} */
    IvyPinchProperties.prototype.wheelZoomFactor;
    /** @type {?|undefined} */
    IvyPinchProperties.prototype.draggableImage;
}
/** @type {?} */
export var IvyPinchDefaultProperties = {
    doubleTap: true,
    doubleTapScale: 2,
    transitionDuration: 200,
    limitZoom: "original image size",
    minScale: 0,
    minPanScale: 1.0001,
    wheel: true,
    wheelZoomFactor: 0.2,
    draggableImage: true,
    listeners: 'auto',
    zoomControlScale: 2
};
var IvyPinch = /** @class */ (function () {
    function IvyPinch(properties) {
        var _this = this;
        this.i = 0;
        this.scale = 1;
        this.initialScale = 1;
        this.startX = 0;
        this.startY = 0;
        this.moveX = 0;
        this.moveY = 0;
        this.initialMoveX = 0;
        this.initialMoveY = 0;
        this.moveXC = 0;
        this.moveYC = 0;
        this.lastTap = 0;
        this.draggingMode = false;
        this.distance = 0;
        this.doubleTapTimeout = 0;
        this.initialDistance = 0;
        this.events = {};
        this.maxHtmlContentScale = 3;
        this.maxScale = 3;
        /* Touchstart */
        this.handleTouchstart = (/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            _this.getElementPosition();
            if (_this.eventType === undefined) {
                _this.getTouchstartPosition(event);
            }
        });
        /* Touchend */
        this.handleTouchend = (/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            /* touchend */
            if (event.type === "touchend") {
                _this.i = 0;
                _this.draggingMode = false;
                /** @type {?} */
                var touches = event.touches;
                // Min scale
                if (_this.scale < 1) {
                    _this.scale = 1;
                }
                // Auto Zoom Out
                if (_this.properties.autoZoomOut && _this.eventType === 'pinch') {
                    _this.scale = 1;
                }
                // Align image
                if (_this.eventType === 'pinch' ||
                    _this.eventType === 'pan' && _this.scale > _this.properties.minPanScale) {
                    _this.alignImage();
                }
                // Update initial values
                if (_this.eventType === 'pinch' ||
                    _this.eventType === 'pan' ||
                    _this.eventType === 'horizontal-swipe' ||
                    _this.eventType === 'vertical-swipe') {
                    _this.updateInitialValues();
                }
                _this.eventType = 'touchend';
                if (touches && touches.length === 0) {
                    _this.eventType = undefined;
                }
            }
            /* mouseup */
            if (event.type === "mouseup") {
                _this.draggingMode = false;
                _this.updateInitialValues();
                _this.eventType = undefined;
            }
        });
        /*
             * Handlers
             */
        this.handlePan = (/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            if (_this.scale < _this.properties.minPanScale || _this.properties.disablePan) {
                return;
            }
            event.preventDefault();
            var _a = _this.getClientPosition(event), clientX = _a.clientX, clientY = _a.clientY;
            if (!_this.eventType) {
                _this.startX = clientX - _this.elementPosition.left;
                _this.startY = clientY - _this.elementPosition.top;
            }
            _this.eventType = 'pan';
            _this.moveX = _this.initialMoveX + (_this.moveLeft(event, 0) - _this.startX);
            _this.moveY = _this.initialMoveY + (_this.moveTop(event, 0) - _this.startY);
            if (_this.properties.limitPan) {
                _this.limitPanY();
                _this.limitPanX();
            }
            /* mousemove */
            if (event.type === "mousemove" && _this.scale > _this.properties.minPanScale) {
                _this.centeringImage();
            }
            _this.transformElement(0);
        });
        this.handleDoubleTap = (/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            _this.toggleZoom(event);
            return;
        });
        this.handlePinch = (/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            event.preventDefault();
            if (_this.eventType === undefined || _this.eventType === 'pinch') {
                /** @type {?} */
                var touches = event.touches;
                if (!_this.eventType) {
                    _this.initialDistance = _this.getDistance(touches);
                    /** @type {?} */
                    var moveLeft0 = _this.moveLeft(event, 0);
                    /** @type {?} */
                    var moveLeft1 = _this.moveLeft(event, 1);
                    /** @type {?} */
                    var moveTop0 = _this.moveTop(event, 0);
                    /** @type {?} */
                    var moveTop1 = _this.moveTop(event, 1);
                    _this.moveXC = ((moveLeft0 + moveLeft1) / 2) - _this.initialMoveX;
                    _this.moveYC = ((moveTop0 + moveTop1) / 2) - _this.initialMoveY;
                }
                _this.eventType = 'pinch';
                _this.distance = _this.getDistance(touches);
                _this.scale = _this.initialScale * (_this.distance / _this.initialDistance);
                _this.moveX = _this.initialMoveX - (((_this.distance / _this.initialDistance) * _this.moveXC) - _this.moveXC);
                _this.moveY = _this.initialMoveY - (((_this.distance / _this.initialDistance) * _this.moveYC) - _this.moveYC);
                _this.handleLimitZoom();
                if (_this.properties.limitPan) {
                    _this.limitPanY();
                    _this.limitPanX();
                }
                _this.transformElement(0);
            }
        });
        this.handleWheel = (/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            event.preventDefault();
            /** @type {?} */
            var wheelZoomFactor = _this.properties.wheelZoomFactor || 0;
            /** @type {?} */
            var zoomFactor = event.deltaY < 0 ? (wheelZoomFactor) : (-wheelZoomFactor);
            /** @type {?} */
            var newScale = _this.initialScale + zoomFactor;
            /* Round value */
            if (newScale < (1 + wheelZoomFactor)) {
                newScale = 1;
            }
            else if (newScale < _this.maxScale && newScale > _this.maxScale - wheelZoomFactor) {
                newScale = _this.maxScale;
            }
            if (newScale < 1 || newScale > _this.maxScale) {
                return;
            }
            if (newScale === _this.scale) {
                return;
            }
            _this.getElementPosition();
            _this.scale = newScale;
            /* Get cursor position over image */
            /** @type {?} */
            var xCenter = (event.clientX - _this.elementPosition.left) - _this.initialMoveX;
            /** @type {?} */
            var yCenter = (event.clientY - _this.elementPosition.top) - _this.initialMoveY;
            _this.setZoom({
                scale: newScale,
                center: [xCenter, yCenter]
            });
        });
        this.handleResize = (/**
         * @param {?} _event
         * @return {?}
         */
        function (_event) {
            _this.setAutoHeight();
        });
        this.element = properties.element;
        this.elementTarget = this.element.querySelector('*').tagName;
        this.parentElement = this.element.parentElement;
        this.properties = Object.assign({}, IvyPinchDefaultProperties, properties);
        this.pollLimitZoom();
        this.touches = new Touches({
            element: properties.element,
            listeners: this.properties.listeners,
            resize: this.properties.autoHeight
        });
        /* Init */
        this.setBasicStyles();
        /*
         * Listeners
         */
        this.touches.on('touchstart', this.handleTouchstart);
        this.touches.on('touchend', this.handleTouchend);
        this.touches.on('mousedown', this.handleTouchstart);
        this.touches.on('mouseup', this.handleTouchend);
        this.touches.on('pan', this.handlePan);
        this.touches.on('mousemove', this.handlePan);
        this.touches.on('pinch', this.handlePinch);
        if (this.properties.wheel) {
            this.touches.on('wheel', this.handleWheel);
        }
        if (this.properties.doubleTap) {
            this.touches.on('double-tap', this.handleDoubleTap);
        }
        if (this.properties.autoHeight) {
            this.touches.on('resize', this.handleResize);
        }
    }
    /**
     * @return {?}
     */
    IvyPinch.prototype.handleLimitZoom = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var limitZoom = this.maxScale;
        /** @type {?} */
        var minScale = this.properties.minScale || 0;
        if (this.scale > limitZoom || this.scale <= minScale) {
            /** @type {?} */
            var imageWidth = this.getImageWidth();
            /** @type {?} */
            var imageHeight = this.getImageHeight();
            /** @type {?} */
            var enlargedImageWidth = imageWidth * this.scale;
            /** @type {?} */
            var enlargedImageHeight = imageHeight * this.scale;
            /** @type {?} */
            var moveXRatio = this.moveX / (enlargedImageWidth - imageWidth);
            /** @type {?} */
            var moveYRatio = this.moveY / (enlargedImageHeight - imageHeight);
            if (this.scale > limitZoom) {
                this.scale = limitZoom;
            }
            if (this.scale <= minScale) {
                this.scale = minScale;
            }
            /** @type {?} */
            var newImageWidth = imageWidth * this.scale;
            /** @type {?} */
            var newImageHeight = imageHeight * this.scale;
            this.moveX = -Math.abs((moveXRatio * (newImageWidth - imageWidth)));
            this.moveY = -Math.abs((-moveYRatio * (newImageHeight - imageHeight)));
        }
    };
    /**
     * @return {?}
     */
    IvyPinch.prototype.getLimitZoom = /**
     * @return {?}
     */
    function () {
        if (this.properties.limitZoom === "original image size") {
            if (this.elementTarget === "IMG") {
                /** @type {?} */
                var img = this.element.getElementsByTagName("img")[0];
                if (img.naturalWidth && img.offsetWidth) {
                    this.maxScale = img.naturalWidth / img.offsetWidth;
                    return this.maxScale;
                }
            }
            else {
                this.maxScale = this.maxHtmlContentScale;
                return this.maxScale;
            }
        }
        else {
            this.maxScale = this.properties.limitZoom || 0;
            return this.maxScale;
        }
    };
    /**
     * @param {?} event
     * @param {?=} index
     * @return {?}
     */
    IvyPinch.prototype.moveLeft = /**
     * @param {?} event
     * @param {?=} index
     * @return {?}
     */
    function (event, index) {
        if (index === void 0) { index = 0; }
        /** @type {?} */
        var clientX = this.getClientPosition(event, index).clientX;
        return clientX - this.elementPosition.left;
    };
    /**
     * @param {?} event
     * @param {?=} index
     * @return {?}
     */
    IvyPinch.prototype.moveTop = /**
     * @param {?} event
     * @param {?=} index
     * @return {?}
     */
    function (event, index) {
        if (index === void 0) { index = 0; }
        /** @type {?} */
        var clientY = this.getClientPosition(event, index).clientY;
        return clientY - this.elementPosition.top;
    };
    /*
     * Detection
     */
    /*
         * Detection
         */
    /**
     * @return {?}
     */
    IvyPinch.prototype.centeringImage = /*
         * Detection
         */
    /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var img = this.element.getElementsByTagName(this.elementTarget)[0];
        /** @type {?} */
        var initialMoveX = this.moveX;
        /** @type {?} */
        var initialMoveY = this.moveY;
        if (this.moveY > 0) {
            this.moveY = 0;
        }
        if (this.moveX > 0) {
            this.moveX = 0;
        }
        if (img) {
            this.limitPanY();
            this.limitPanX();
        }
        if (img && this.scale < 1) {
            if (this.moveX < this.element.offsetWidth * (1 - this.scale)) {
                this.moveX = this.element.offsetWidth * (1 - this.scale);
            }
        }
        return initialMoveX !== this.moveX || initialMoveY !== this.moveY;
    };
    /**
     * @return {?}
     */
    IvyPinch.prototype.limitPanY = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var imgHeight = this.getImageHeight();
        /** @type {?} */
        var scaledImgHeight = imgHeight * this.scale;
        /** @type {?} */
        var parentHeight = this.parentElement.offsetHeight;
        /** @type {?} */
        var elementHeight = this.element.offsetHeight;
        if (scaledImgHeight < parentHeight) {
            this.moveY = (parentHeight - elementHeight * this.scale) / 2;
        }
        else {
            /** @type {?} */
            var imgOffsetTop = ((imgHeight - elementHeight) * this.scale) / 2;
            if (this.moveY > imgOffsetTop) {
                this.moveY = imgOffsetTop;
            }
            else if ((scaledImgHeight + Math.abs(imgOffsetTop) - parentHeight) + this.moveY < 0) {
                this.moveY = -(scaledImgHeight + Math.abs(imgOffsetTop) - parentHeight);
            }
        }
    };
    /**
     * @return {?}
     */
    IvyPinch.prototype.limitPanX = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var imgWidth = this.getImageWidth();
        /** @type {?} */
        var scaledImgWidth = imgWidth * this.scale;
        /** @type {?} */
        var parentWidth = this.parentElement.offsetWidth;
        /** @type {?} */
        var elementWidth = this.element.offsetWidth;
        if (scaledImgWidth < parentWidth) {
            this.moveX = (parentWidth - elementWidth * this.scale) / 2;
        }
        else {
            /** @type {?} */
            var imgOffsetLeft = ((imgWidth - elementWidth) * this.scale) / 2;
            if (this.moveX > imgOffsetLeft) {
                this.moveX = imgOffsetLeft;
            }
            else if ((scaledImgWidth + Math.abs(imgOffsetLeft) - parentWidth) + this.moveX < 0) {
                this.moveX = -(imgWidth * this.scale + Math.abs(imgOffsetLeft) - parentWidth);
            }
        }
    };
    /**
     * @return {?}
     */
    IvyPinch.prototype.setBasicStyles = /**
     * @return {?}
     */
    function () {
        this.element.style.display = 'flex';
        this.element.style.alignItems = 'center';
        this.element.style.justifyContent = 'center';
        this.element.style.transformOrigin = '0 0';
        this.setImageSize();
        this.setDraggableImage();
    };
    /**
     * @return {?}
     */
    IvyPinch.prototype.removeBasicStyles = /**
     * @return {?}
     */
    function () {
        this.element.style.display = '';
        this.element.style.alignItems = '';
        this.element.style.justifyContent = '';
        this.element.style.transformOrigin = '';
        this.removeImageSize();
        this.removeDraggableImage();
    };
    /**
     * @return {?}
     */
    IvyPinch.prototype.setDraggableImage = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var imgElement = this.getImageElement();
        if (imgElement) {
            imgElement.draggable = this.properties.draggableImage;
        }
    };
    /**
     * @return {?}
     */
    IvyPinch.prototype.removeDraggableImage = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var imgElement = this.getImageElement();
        if (imgElement) {
            imgElement.draggable = true;
        }
    };
    /**
     * @return {?}
     */
    IvyPinch.prototype.setImageSize = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var imgElement = this.element.getElementsByTagName(this.elementTarget);
        if (imgElement.length) {
            imgElement[0].style.maxWidth = '100%';
            imgElement[0].style.maxHeight = '100%';
            this.setAutoHeight();
        }
    };
    /**
     * @return {?}
     */
    IvyPinch.prototype.setAutoHeight = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var imgElement = this.element.getElementsByTagName(this.elementTarget);
        if (!this.properties.autoHeight || !imgElement.length) {
            return;
        }
        /** @type {?} */
        var imgNaturalWidth = imgElement[0].getAttribute("width");
        /** @type {?} */
        var imgNaturalHeight = imgElement[0].getAttribute("height");
        /** @type {?} */
        var sizeRatio = imgNaturalWidth / imgNaturalHeight;
        /** @type {?} */
        var parentWidth = this.parentElement.offsetWidth;
        imgElement[0].style.maxHeight = parentWidth / sizeRatio + "px";
    };
    /**
     * @return {?}
     */
    IvyPinch.prototype.removeImageSize = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var imgElement = this.element.getElementsByTagName(this.elementTarget);
        if (imgElement.length) {
            imgElement[0].style.maxWidth = '';
            imgElement[0].style.maxHeight = '';
        }
    };
    /**
     * @return {?}
     */
    IvyPinch.prototype.getElementPosition = /**
     * @return {?}
     */
    function () {
        this.elementPosition = this.element.parentElement.getBoundingClientRect();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    IvyPinch.prototype.getTouchstartPosition = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        var _a = this.getClientPosition(event), clientX = _a.clientX, clientY = _a.clientY;
        this.startX = clientX - this.elementPosition.left;
        this.startY = clientY - this.elementPosition.top;
    };
    /**
     * @param {?} event
     * @param {?=} index
     * @return {?}
     */
    IvyPinch.prototype.getClientPosition = /**
     * @param {?} event
     * @param {?=} index
     * @return {?}
     */
    function (event, index) {
        if (index === void 0) { index = 0; }
        /** @type {?} */
        var clientX;
        /** @type {?} */
        var clientY;
        if (event.type === "touchstart" || event.type === "touchmove") {
            clientX = event.touches[index].clientX;
            clientY = event.touches[index].clientY;
        }
        if (event.type === "mousedown" || event.type === "mousemove") {
            clientX = event.clientX;
            clientY = event.clientY;
        }
        return {
            clientX: clientX,
            clientY: clientY
        };
    };
    /**
     * @return {?}
     */
    IvyPinch.prototype.resetScale = /**
     * @return {?}
     */
    function () {
        this.scale = 1;
        this.moveX = 0;
        this.moveY = 0;
        this.updateInitialValues();
        this.transformElement(this.properties.transitionDuration);
    };
    /**
     * @return {?}
     */
    IvyPinch.prototype.updateInitialValues = /**
     * @return {?}
     */
    function () {
        this.initialScale = this.scale;
        this.initialMoveX = this.moveX;
        this.initialMoveY = this.moveY;
    };
    /**
     * @param {?} touches
     * @return {?}
     */
    IvyPinch.prototype.getDistance = /**
     * @param {?} touches
     * @return {?}
     */
    function (touches) {
        return Math.sqrt(Math.pow(touches[0].pageX - touches[1].pageX, 2) + Math.pow(touches[0].pageY - touches[1].pageY, 2));
    };
    /**
     * @return {?}
     */
    IvyPinch.prototype.getImageHeight = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var img = this.element.getElementsByTagName(this.elementTarget)[0];
        return img.offsetHeight;
    };
    /**
     * @return {?}
     */
    IvyPinch.prototype.getImageWidth = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var img = this.element.getElementsByTagName(this.elementTarget)[0];
        return img.offsetWidth;
    };
    /**
     * @param {?} duration
     * @return {?}
     */
    IvyPinch.prototype.transformElement = /**
     * @param {?} duration
     * @return {?}
     */
    function (duration) {
        this.element.style.transition = "all " + duration + "ms";
        this.element.style.transform = "matrix(" + Number(this.scale) + ", 0, 0, " + Number(this.scale) + ", " + Number(this.moveX) + ", " + Number(this.moveY) + ")";
    };
    /**
     * @return {?}
     */
    IvyPinch.prototype.isTouchScreen = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var prefixes = ' -webkit- -moz- -o- -ms- '.split(' ');
        if (('ontouchstart' in window)) {
            return true;
        }
        // include the 'heartz' as a way to have a non matching MQ to help terminate the join
        // https://git.io/vznFH
        /** @type {?} */
        var query = ['(', prefixes.join('touch-enabled),('), 'heartz', ')'].join('');
        return this.getMatchMedia(query);
    };
    /**
     * @param {?} query
     * @return {?}
     */
    IvyPinch.prototype.getMatchMedia = /**
     * @param {?} query
     * @return {?}
     */
    function (query) {
        return window.matchMedia(query).matches;
    };
    /**
     * @return {?}
     */
    IvyPinch.prototype.isDragging = /**
     * @return {?}
     */
    function () {
        if (this.properties.disablePan) {
            return false;
        }
        /** @type {?} */
        var imgHeight = this.getImageHeight();
        /** @type {?} */
        var imgWidth = this.getImageWidth();
        if (this.scale > 1) {
            return imgHeight * this.scale > this.parentElement.offsetHeight ||
                imgWidth * this.scale > this.parentElement.offsetWidth;
        }
        if (this.scale === 1) {
            return imgHeight > this.parentElement.offsetHeight ||
                imgWidth > this.parentElement.offsetWidth;
        }
    };
    /**
     * @return {?}
     */
    IvyPinch.prototype.pollLimitZoom = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var poll = setInterval((/**
         * @return {?}
         */
        function () {
            if (_this.getLimitZoom()) {
                clearInterval(poll);
            }
        }), 10);
    };
    /**
     * @return {?}
     */
    IvyPinch.prototype.getImageElement = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var imgElement = this.element.getElementsByTagName(this.elementTarget);
        if (imgElement.length) {
            return imgElement[0];
        }
    };
    /**
     * @param {?=} event
     * @return {?}
     */
    IvyPinch.prototype.toggleZoom = /**
     * @param {?=} event
     * @return {?}
     */
    function (event) {
        if (event === void 0) { event = false; }
        if (this.initialScale === 1) {
            if (event && event.changedTouches) {
                if (this.properties.doubleTapScale === undefined) {
                    return;
                }
                /** @type {?} */
                var changedTouches = event.changedTouches;
                this.scale = this.initialScale * this.properties.doubleTapScale;
                this.moveX = this.initialMoveX - (changedTouches[0].clientX - this.elementPosition.left) * (this.properties.doubleTapScale - 1);
                this.moveY = this.initialMoveY - (changedTouches[0].clientY - this.elementPosition.top) * (this.properties.doubleTapScale - 1);
            }
            else {
                /** @type {?} */
                var zoomControlScale = this.properties.zoomControlScale || 0;
                this.scale = this.initialScale * (zoomControlScale + 1);
                this.moveX = this.initialMoveX - this.element.offsetWidth * (this.scale - 1) / 2;
                this.moveY = this.initialMoveY - this.element.offsetHeight * (this.scale - 1) / 2;
            }
            this.centeringImage();
            this.updateInitialValues();
            this.transformElement(this.properties.transitionDuration);
        }
        else {
            this.resetScale();
        }
    };
    /**
     * @param {?} properties
     * @return {?}
     */
    IvyPinch.prototype.setZoom = /**
     * @param {?} properties
     * @return {?}
     */
    function (properties) {
        this.scale = properties.scale;
        /** @type {?} */
        var xCenter;
        /** @type {?} */
        var yCenter;
        /** @type {?} */
        var visibleAreaWidth = this.element.offsetWidth;
        /** @type {?} */
        var visibleAreaHeight = this.element.offsetHeight;
        /** @type {?} */
        var scalingPercent = (visibleAreaWidth * this.scale) / (visibleAreaWidth * this.initialScale);
        if (properties.center) {
            xCenter = properties.center[0];
            yCenter = properties.center[1];
        }
        else {
            xCenter = visibleAreaWidth / 2 - this.initialMoveX;
            yCenter = visibleAreaHeight / 2 - this.initialMoveY;
        }
        this.moveX = this.initialMoveX - ((scalingPercent * xCenter) - xCenter);
        this.moveY = this.initialMoveY - ((scalingPercent * yCenter) - yCenter);
        this.centeringImage();
        this.updateInitialValues();
        this.transformElement(this.properties.transitionDuration);
    };
    /**
     * @return {?}
     */
    IvyPinch.prototype.alignImage = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var isMoveChanged = this.centeringImage();
        if (isMoveChanged) {
            this.updateInitialValues();
            this.transformElement(this.properties.transitionDuration);
        }
    };
    /**
     * @return {?}
     */
    IvyPinch.prototype.destroy = /**
     * @return {?}
     */
    function () {
        this.removeBasicStyles();
        this.touches.destroy();
    };
    return IvyPinch;
}());
export { IvyPinch };
if (false) {
    /** @type {?} */
    IvyPinch.prototype.properties;
    /** @type {?} */
    IvyPinch.prototype.touches;
    /** @type {?} */
    IvyPinch.prototype.element;
    /** @type {?} */
    IvyPinch.prototype.elementTarget;
    /** @type {?} */
    IvyPinch.prototype.parentElement;
    /** @type {?} */
    IvyPinch.prototype.i;
    /** @type {?} */
    IvyPinch.prototype.scale;
    /** @type {?} */
    IvyPinch.prototype.initialScale;
    /** @type {?} */
    IvyPinch.prototype.elementPosition;
    /** @type {?} */
    IvyPinch.prototype.eventType;
    /** @type {?} */
    IvyPinch.prototype.startX;
    /** @type {?} */
    IvyPinch.prototype.startY;
    /** @type {?} */
    IvyPinch.prototype.moveX;
    /** @type {?} */
    IvyPinch.prototype.moveY;
    /** @type {?} */
    IvyPinch.prototype.initialMoveX;
    /** @type {?} */
    IvyPinch.prototype.initialMoveY;
    /** @type {?} */
    IvyPinch.prototype.moveXC;
    /** @type {?} */
    IvyPinch.prototype.moveYC;
    /** @type {?} */
    IvyPinch.prototype.lastTap;
    /** @type {?} */
    IvyPinch.prototype.draggingMode;
    /** @type {?} */
    IvyPinch.prototype.distance;
    /** @type {?} */
    IvyPinch.prototype.doubleTapTimeout;
    /** @type {?} */
    IvyPinch.prototype.initialDistance;
    /** @type {?} */
    IvyPinch.prototype.events;
    /** @type {?} */
    IvyPinch.prototype.maxHtmlContentScale;
    /** @type {?} */
    IvyPinch.prototype.maxScale;
    /** @type {?} */
    IvyPinch.prototype.handleTouchstart;
    /** @type {?} */
    IvyPinch.prototype.handleTouchend;
    /** @type {?} */
    IvyPinch.prototype.handlePan;
    /** @type {?} */
    IvyPinch.prototype.handleDoubleTap;
    /** @type {?} */
    IvyPinch.prototype.handlePinch;
    /** @type {?} */
    IvyPinch.prototype.handleWheel;
    /** @type {?} */
    IvyPinch.prototype.handleResize;
}
