// ngx-pinch-zoom/esm5/lib/pinch-zoom.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pinch-zoom.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { PinchZoomComponent } from './pinch-zoom.component';
var PinchZoomModule = /** @class */ (function () {
    function PinchZoomModule() {
    }
    PinchZoomModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        PinchZoomComponent
                    ],
                    imports: [
                        CommonModule
                    ],
                    exports: [
                        PinchZoomComponent
                    ],
                    providers: [],
                    bootstrap: [],
                    entryComponents: [
                        PinchZoomComponent
                    ]
                },] }
    ];
    return PinchZoomModule;
}());
export { PinchZoomModule };
